hail/src/main/scala/is/hail/expr/ir/functions/StringFunctions.scala

```diff
@@ -133,50 +135,82 @@ object StringFunctions extends RegistryFunctions {
       unwrapReturn(r, rt)(str)
     }
 
-    registerCodeWithMissingness("showStr", tv("T"), TInt32, TString, null) { case (r, rt, (aT, a), (_, trunc)) =>
+    registerCodeWithMissingness("showStr", tv("T"), TInt32, TString, {
+      (_: PType, truncType: PType) => PCanonicalString(truncType.required)
+    }) { case (r, rt, (aT, a), (_, trunc)) =>
       val annotation = Code(a.setup, a.m).muxAny(Code._null(boxedTypeInfo(aT)), boxArg(r, aT)(a.v))
       val str = r.mb.getType(aT.virtualType).invoke[Any, Int, String]("showStr", annotation, trunc.value[Int])
       EmitCode(trunc.setup, trunc.m, PCode(rt, unwrapReturn(r, rt)(str)))
     }
 
-    registerCodeWithMissingness("json", tv("T"), TString, null) { case (r, rt, (aT, a)) =>
+    registerCodeWithMissingness("json", tv("T"), TString, (_: PType) => PCanonicalString(true)) { case (r, rt, (aT, a)) =>
       val bti = boxedTypeInfo(aT)
       val annotation = Code(a.setup, a.m).muxAny(Code._null(bti), boxArg(r, aT)(a.v))
       val json = r.mb.getType(aT.virtualType).invoke[Any, JValue]("toJSON", annotation)
       val str = Code.invokeScalaObject[JValue, String](JsonMethods.getClass, "compact", json)
       EmitCode(Code._empty, false, PCode(rt, unwrapReturn(r, rt)(str)))
     }
 
-    registerWrappedScalaFunction("reverse", TString, TString, null)(thisClass,"reverse")
-    registerWrappedScalaFunction("upper", TString, TString, null)(thisClass,"upper")
-    registerWrappedScalaFunction("lower", TString, TString, null)(thisClass,"lower")
-    registerWrappedScalaFunction("strip", TString, TString, null)(thisClass,"strip")
-    registerWrappedScalaFunction("contains", TString, TString, TBoolean, null)(thisClass, "contains")
-    registerWrappedScalaFunction("translate", TString, TDict(TString, TString), TString, null)(thisClass, "translate")
-    registerWrappedScalaFunction("startswith", TString, TString, TBoolean, null)(thisClass, "startswith")
-    registerWrappedScalaFunction("endswith", TString, TString, TBoolean, null)(thisClass, "endswith")
-
-    registerWrappedScalaFunction("regexMatch", TString, TString, TBoolean, null)(thisClass, "regexMatch")
-
-    registerWrappedScalaFunction("concat", TString, TString, TString, null)(thisClass, "concat")
-
-    registerWrappedScalaFunction("split", TString, TString, TArray(TString), null)(thisClass, "split")
+    registerWrappedScalaFunction("reverse", TString, TString, (pt: PType) => pt)(thisClass,"reverse")
+    registerWrappedScalaFunction("upper", TString, TString, (pt: PType) => pt)(thisClass,"upper")
+    registerWrappedScalaFunction("lower", TString, TString, (pt: PType) => pt)(thisClass,"lower")
+    registerWrappedScalaFunction("strip", TString, TString, (pt: PType) => pt)(thisClass,"strip")
+    registerWrappedScalaFunction("contains", TString, TString, TBoolean, {
+      case (strPT: PType, substrPT: PType) => PBoolean(strPT.required && substrPT.required)
```

Comment:
don't do top-level requiredness here and below, for the same reason

Reply:
kk